kpatch-build/create-diff-object.c

```diff
@@ -3659,29 +3660,51 @@ static void kpatch_alloc_mcount_sections(struct kpatch_elf *kelf, struct kpatch_
 {
 	int nr;
 	struct symbol *sym;
+	int text_idx = 0;
 
 	nr = 0;
-	list_for_each_entry(sym, &kelfout->symbols, list)
+	list_for_each_entry(sym, &kelfout->symbols, list) {
 		if (sym->type == STT_FUNC && sym->status != SAME &&
-		    sym->has_func_profiling)
+		    sym->has_func_profiling) {
+			text_idx = sym->sec->index;
 			nr++;
+		}
+	}
 
 	/* create text/rela section pair */
 	switch(kelf->arch) {
 	case AARCH64: {
-		struct section *sec, *tmp;
-
-		sec = create_section_pair(kelfout, "__patchable_function_entries", sizeof(void*), nr);
+		struct section *sec;
+		int entries = multi_pfe ? 1 : nr;
+		int copies = multi_pfe ? nr : 1;
+		int flags = 0, rflags = 0;
 
 		/*
 		 * Depending on the compiler the __patchable_function_entries section
 		 * can be ordered or not, copy this flag to the section we created to
 		 * avoid:
 		 * ld: __patchable_function_entries has both ordered [...] and unordered [...] sections
 		 */
-		tmp = find_section_by_name(&kelf->sections, "__patchable_function_entries");
-		sec->sh.sh_flags |= (tmp->sh.sh_flags & SHF_LINK_ORDER);
-		sec->sh.sh_link = 1;
+		sec = find_section_by_name(&kelf->sections, "__patchable_function_entries");
+		if (sec) {
+			flags = (sec->sh.sh_flags & (SHF_LINK_ORDER|SHF_WRITE));
+			if (sec->rela)
+				rflags = (sec->rela->sh.sh_flags & (SHF_LINK_ORDER|SHF_WRITE));
+		}
+
+		for (nr = 0; nr < copies; nr++) {
+			sec = create_section_pair(kelfout,
+				"__patchable_function_entries",
+				sizeof(void*), entries);
+
+			sec->sh.sh_flags |= flags;
+			if (sec->rela)
+				sec->rela->sh.sh_flags |= rflags;
+			if (multi_pfe)
+				sec->sh.sh_link = 0; /* set later */
+			else
+				sec->sh.sh_link = text_idx;
```

Comment:
I saw the following error with gcc12 for integration test (data-new etc.):

> ld: __patchable_function_entries has both ordered and unordered sections

Bisect points this commit. With additional log messages, I found text_idx is 0 and it causes the problem.
Should we set sh_link later like when multi_pfe is true or set 1 unconditionally just like before this commit?
(I'm looking this comment...: https://github.com/dynup/kpatch/pull/1236#issuecomment-1040197012)

Reply:
I think the issue was that multi_pfe was getting set too late.
Next push should resolve that

Reply:
Hi @swine, I'm still seeing
```
ld: __patchable_function_entries has both ordered and unordered sections
```
with gcc12 and your latest push (2023.1.16, d01cbc2), set sh_link to 1 unconditionally fixed the issue for me.